File: OpenWA/sdk/java/src/main/java/com/rmyndharis/openwa/resources/TemplatesResource.java
package com.rmyndharis.openwa.resources;

import static com.rmyndharis.openwa.http.Http.encodeSegment;

import com.rmyndharis.openwa.OpenWAClient;
import com.rmyndharis.openwa.http.HttpMethod;
import com.rmyndharis.openwa.model.CreateTemplateRequest;
import com.rmyndharis.openwa.model.TemplateRecord;
import com.rmyndharis.openwa.model.UpdateTemplateRequest;
import java.util.List;

/** Templates resource — stored message templates with {@code {{variable}}} placeholders. */
public final class TemplatesResource {
    private final OpenWAClient client;

    public TemplatesResource(OpenWAClient client) {
        this.client = client;
    }

    /** List all templates for a session. */
    public List<TemplateRecord> list(String sessionId) {
        return client.requestList(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/templates",
            null,
            null,
            TemplateRecord.class);
    }

    /** Get a single template by id. */
    public TemplateRecord get(String sessionId, String id) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/templates/" + encodeSegment(id),
            null,
            null,
            TemplateRecord.class);
    }

    /** Create a new template. */
    public TemplateRecord create(String sessionId, CreateTemplateRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/templates",
            null,
            body,
            TemplateRecord.class);
    }

    /** Update a template. */
    public TemplateRecord update(String sessionId, String id, UpdateTemplateRequest body) {
        return client.request(
            HttpMethod.PUT,
            "/api/sessions/" + encodeSegment(sessionId) + "/templates/" + encodeSegment(id),
            null,
            body,
            TemplateRecord.class);
    }

    /** Delete a template. */
    public void delete(String sessionId, String id) {
        client.requestVoid(
            HttpMethod.DELETE,
            "/api/sessions/" + encodeSegment(sessionId) + "/templates/" + encodeSegment(id),
            null,
            null);
    }
}

File: OpenWA/sdk/php/src/Resources/TemplatesResource.php
<?php

declare(strict_types=1);

namespace OpenWA\Resources;

use OpenWA\Http\HttpExecutor;

/**
 * Templates resource — stored message templates with {{variable}} placeholders.
 *
 * Backed by src/modules/template/template.controller.ts
 * (@Controller('sessions/:sessionId/templates')).
 */
class TemplatesResource
{
    private HttpExecutor $http;

    public function __construct(HttpExecutor $http)
    {
        $this->http = $http;
    }

    /** @return array<int,array<string,mixed>> */
    public function list(string $sessionId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/templates") ?? [];
    }

    /** @return array<string,mixed> */
    public function get(string $sessionId, string $templateId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/templates/{$this->http->encodeSegment($templateId)}");
    }

    /**
     * Create a template. Requires an OPERATOR-level key.
     *
     * @param array<string,mixed> $body  Must contain 'name' and 'body'; 'header'/'footer' optional.
     * @return array<string,mixed>
     */
    public function create(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/templates", [], $body);
    }

    /**
     * Update a template. Requires an OPERATOR-level key.
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function update(string $sessionId, string $templateId, array $body): array
    {
        return $this->http->request('PUT', "/api/sessions/{$this->http->encodeSegment($sessionId)}/templates/{$this->http->encodeSegment($templateId)}", [], $body);
    }

    /** Delete a template. Requires an OPERATOR-level key. */
    public function delete(string $sessionId, string $templateId): void
    {
        $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/templates/{$this->http->encodeSegment($templateId)}");
    }
}
